File: android/src/main/java/io/sariska/sariska_media_flutter_sdk/TrackPlugin.java
package io.sariska.sariska_media_flutter_sdk;

import androidx.annotation.NonNull;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class TrackPlugin implements MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private MethodChannel methodChannel;
    private TrackManager manager = new TrackManager();

    public  TrackPlugin(BinaryMessenger messenger){
        methodChannel = new MethodChannel(messenger, "track");
        methodChannel.setMethodCallHandler(this);
    }


    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Method[] methods = manager.getClass().getDeclaredMethods();
        System.out.println("Calling track methods");
        System.out.println(call.method);
        for (Method method : methods) {
            if (method.getName().equals(call.method)) {
                try {
                    List<Object> parameters = new ArrayList<>();
                    Map<?, ?> arguments = call.arguments();
                    if (arguments != null) {
                        parameters.add(arguments);
                    }
                    method.invoke(manager, parameters.toArray());
                    return;
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
        result.notImplemented();
    }

    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {

    }

    @Override
    public void onCancel(Object arguments) {

    }
}

File: ios/Classes/TrackPlugin.swift
//
//  TrackPlugin.swift
//  sariska_media_flutter_sdk
//
//  Created by Dipak Sisodiya on 24/11/23.
//

import Foundation
import sariska
import Flutter

public class TrackPlugin: NSObject, FlutterPlugin, FlutterStreamHandler{
    
    var manager: TrackManager?
    
    private var eventSink: FlutterEventSink? = nil
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "track", binaryMessenger: registrar.messenger())
        let instance = TrackPlugin()
        
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if(manager == nil){
            manager = TrackManager()
        }
        
        if let params = call.arguments {
            // if there are parameters
            let selectorString = call.method + ":"
            let selector = Selector(selectorString)
            if manager?.responds(to: selector) != nil{
                manager?.perform(selector, with: params)
            }
        } else {
            // if there are no parameters
            let selectorString = call.method
            let selector = Selector(selectorString)
            if manager?.responds(to: selector) != nil{
                manager?.perform(selector)
            }
        }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}
